File: backend/src/main/java/com/tojaeung/blog/comment/domain/Comment.java
package com.tojaeung.blog.comment.domain;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.tojaeung.blog.post.domain.Post;
import lombok.*;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import javax.persistence.*;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

@Builder
@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor
@EntityListeners(AuditingEntityListener.class)
public class Comment {
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    @Column(name = "comment_id")
    private Long id;

    @Column
    private String author;

    @Column
    private String content;

    @Column
    private Boolean isAdmin;

    @JsonIgnore
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "post_id")
    private Post post;

    @JsonIgnore
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent_id")
    private Comment parent;

    @Builder.Default
    @OneToMany(mappedBy = "parent", orphanRemoval = true)
    private List<Comment> children = new ArrayList<>();

    @CreatedDate
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDate createdAt;
    @LastModifiedDate
    @Column(name = "last_modified_at", nullable = false)
    private LocalDate lastModifiedAt;
}

File: frontend/src/interfaces/comment.ts
export type IComment = {
  id: number;
  author: string;
  content: string;
  isAdmin: boolean;
  postId: number;
  parentId: number;
  children: IComment[];
  createdAt: string;
  lastModifiedAt: string;
};

export type INewComment = {
  postId: number;
  author: string;
  content: string;
  isAdmin: boolean;
};
export type INewChildComment = {
  postId: number;
  author: string;
  content: string;
  isAdmin: boolean;
  parentId: number;
};
